### ISSUE TITLE
False Positive: Casts that are actually necessary or benchmark-excluded are still judged unnecessary.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6539]
# False Positive: Casts that are actually necessary or benchmark-excluded are still judged unnecessary.

Version
7.22.0

## Checker
- Checker id: `category/java/codestyle.xml/UnnecessaryCast`
- Checker description: This rule detects unnecessary cast expressions where the cast is redundant because the operand's type already satisfies the context's requirements or the cast does not affect the semantics.

## Description of the false positive
`diff/output` marks 3 false positives for `category/java/codestyle.xml/UnnecessaryCast`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The reports read more like over-eager matching than strong bug findings.

### `NegCase2.java`
Source: `output/pmd/UnnecessaryCast/src/main/java/scensct/core/neg/NegCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Cast expression with lambda operand in standalone assignment should not be flagged as unnecessary.

```java
// Cast expression with lambda operand in standalone assignment should not be flagged as unnecessary.
package scensct.core.neg;

public class NegCase2 {
    interface MyFunc {
        void run();
    }
    
    public static void main(String[] args) {
        // Cast provides explicit target typing for lambda in variable declaration
        MyFunc func = (MyFunc) () -> { System.out.println("Hello"); }; // Scenario 2: cast gives target type // [REPORTED LINE]
        func.run();
    }
}
```

### `NegCase5.java`
Source: `output/pmd/UnnecessaryCast/src/main/java/scensct/core/neg/NegCase5.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Cast expression with lambda operand in return statement should not be flagged as unnecessary.

```java
// Cast expression with lambda operand in return statement should not be flagged as unnecessary.
package scensct.core.neg;

public class NegCase5 {
    interface Provider {
        int get();
    }
    
    static Provider getProvider() {
        // Cast to Provider in return statement with lambda operand
        return (Provider) () -> { return 42; }; // Scenario 5: cast not trivially unnecessary // [REPORTED LINE]
    }
    
    public static void main(String[] args) {
        System.out.println(getProvider().get());
    }
}
```

### `NegCase2.java`
Source: `output/sonarqube/RedundantTypeCastCheck/src/main/java/scensct/core/neg/NegCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Cast as immediate operand of getClass() should not be flagged as unnecessary.

```java
// Cast as immediate operand of getClass() should not be flagged as unnecessary.
package scensct.core.neg;

public class NegCase2 {
    public static void main(String[] args) {
        Object obj = "test";
        // Cast is directly inside member select for getClass()
        Class<?> clazz = ((String) obj).getClass();
        System.out.println(clazz);
    }
}
```

## Cause analysis
The rule is documented as follows: This rule detects unnecessary cast expressions where the cast is redundant because the operand's type already satisfies the context's requirements or the cast does not affect the semantics.

 The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `category/java/codestyle.xml/UnnecessaryCast`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
